File: __ShinyApp/build_data.R
#!/usr/bin/env Rscript
# =============================================================================
# build_data.R  --  regenerate the Shiny app's derived data from the canonical
# repo files. data/ holds only:
#   * evom1_traits_long.csv  (DERIVED: melted from ____EvoM1_TraitTable/*.xlsx)
#   * source_manifest.csv    (DERIVED: source-table catalogue + citations)
#   * volumes_long.csv, cellcounts_long.csv  (small fallback copies)
#
# The app reads everything from GitHub at runtime; these files are the fallback
# plus the two things that don't exist anywhere else. Re-run whenever the merge
# or trait tables change, then commit + push:
#
#     Rscript __ShinyApp/build_data.R
#
# Requires: readxl  (install.packages("readxl"))
# =============================================================================

suppressWarnings(suppressMessages(library(readxl)))

# ---- locate repo root (parent of this script's folder) ----------------------
args     <- commandArgs(trailingOnly = FALSE)
file_arg <- sub("^--file=", "", args[grep("^--file=", args)])
app_dir  <- if (length(file_arg)) dirname(normalizePath(file_arg)) else getwd()
repo     <- normalizePath(file.path(app_dir, ".."))
out      <- file.path(app_dir, "data")
dir.create(out, recursive = TRUE, showWarnings = FALSE)
message("repo: ", repo)
message("out:  ", out)

trim <- function(x) trimws(as.character(x))

# ---- 1. fallback copies of the two compiled long tables ---------------------
file.copy(file.path(repo, "__merging_volumes", "volumes_long.csv"),
          file.path(out, "volumes_long.csv"), overwrite = TRUE)
file.copy(file.path(repo, "__merging_cellcounts", "cellcounts_long.csv"),
          file.path(out, "cellcounts_long.csv"), overwrite = TRUE)

# ---- 2. melt the EvoM1 trait tables -> evom1_traits_long.csv -----------------
TT <- file.path(repo, "____EvoM1_TraitTable")
# Label = the publication each file's values come from (shown as the app's
# Source column). Per-cell "<col>_Ref"/"_Source" columns override this where the
# source table records a value-specific primary reference (e.g. CST fibre data).
trait_files <- c(
  "dexterity_corticospinaltract.xlsx" = "Heffner & Masterton 1975 (dexterity)",
  "corticospinaltract_etc.xlsx"       = "Iwaniuk et al. 1999 (corticospinal tract & ecology)",
  "glia_gyrification.xlsx"            = "Lewitus et al. 2014 (glia, gyrification & life history)",
  "interlaminar_astrocytes.xlsx"      = "Falcone et al. 2019 (interlaminar astrocytes)",
  "locomotion.xlsx"                   = "Granatosky 2018 (locomotion)",
  "gait.xlsx"                         = "Wimberly et al. 2021 (walking gait)",
  "manipulation.xlsx"                 = "Heldstab et al. 2016 (manipulation)",
  "handedness.xlsx"                   = "Caspar et al. 2022 (handedness)",
  "diet_foraging.xlsx"               = "Wilman et al. 2014 (EltonTraits diet & foraging)",
  "v1_synapses_karl.xlsx"            = "Karl et al. 2024 (V1 synapses & mitochondria)",
  "vocal_repertoire_schniter.xlsx"   = "Schniter & Peñaherrera-Aguirre 2026 (vocal repertoire size)",
  "vocal_repertoire_manyprimates.xlsx" = "ManyPrimates 2022 (vocal repertoire size)",
  "sleep.xlsx"                       = "Eagleman & Vaughn 2021 / Herculano-Houzel 2015 (sleep)"
)
id_cols  <- c("species_sci", "Species", "Animal", "Species Generic Name")
suffixes <- c("_Source", " Source", "_Ref", " Ref", "_ref", " ref")
is_src   <- function(c) any(endsWith(c, suffixes))
base_of  <- function(c) { for (s in suffixes) if (endsWith(c, s))
                            return(trimws(substr(c, 1, nchar(c) - nchar(s)))); c }
# Species key: prefer the harmonised binomial (species_sci) over the paper's
# printed name, then normalise (drop *, underscores->space) so trait species
# match the cell-count / volume species and can be correlated.
clean_sp <- function(x) {
  x <- gsub("\\*", "", trim(x)); x <- gsub("_", " ", x)
  trimws(gsub("\\s+", " ", x))
}

trait_rows <- vector("list", 0L)
for (i in seq_along(trait_files)) {
  fn <- names(trait_files)[i]; label <- unname(trait_files[i])
  d  <- as.data.frame(read_excel(file.path(TT, fn), sheet = "Sheet1",
                                 col_types = "text", .name_repair = "minimal"),
                      stringsAsFactors = FALSE, check.names = FALSE)
  cols <- names(d)
  src_map <- list()
  for (cn in cols) if (nzchar(cn) && is_src(cn)) src_map[[base_of(cn)]] <- cn
  has_sci <- "species_sci" %in% cols
  sp_col  <- if ("Species" %in% cols) "Species" else cols[1]
  for (r in seq_len(nrow(d))) {
    sci <- if (has_sci) d[["species_sci"]][r] else NA
    sp  <- if (!is.na(sci) && nzchar(trim(sci)) && tolower(trim(sci)) != "none")
             sci else d[[sp_col]][r]
    if (is.na(sp) || !nzchar(trim(sp)) || tolower(trim(sp)) == "none") next
    sp <- clean_sp(sp)
    for (cn in cols) {
      if (!nzchar(cn) || cn %in% id_cols || is_src(cn)) next
      v <- d[[cn]][r]
      if (is.na(v)) next
      vs <- trim(v)
      if (!nzchar(vs) || tolower(vs) %in% c("na", "nan", "none", "-")) next
      src <- label
      sc  <- src_map[[cn]]
      if (!is.null(sc)) {
        sv <- d[[sc]][r]
        if (!is.na(sv) && nzchar(trim(sv))) src <- trim(sv)
      }
      trait_rows[[length(trait_rows) + 1L]] <-
        data.frame(Species = sp, Variable = cn, Value = vs, Source = src,
                   stringsAsFactors = FALSE)
    }
  }
}
traits <- unique(do.call(rbind, trait_rows))
write.csv(traits, file.path(out, "evom1_traits_long.csv"),
          row.names = FALSE, fileEncoding = "UTF-8")
message("evom1 traits rows: ", nrow(traits))

# ---- 3. index the public source tables (served from GitHub, not copied) -----
pub  <- file.path(repo, "__Public", "comparative-data")
tsvs <- sort(list.files(pub, pattern = "\\.tsv$"))
mds  <- list.files(pub, pattern = "\\.ReadMe\\.md$")
message("source tables indexed (served from GitHub): ", length(tsvs))

# ---- 4. build source_manifest.csv (filenames + citations from __ReadMe.xlsx) -
readme <- as.data.frame(read_excel(file.path(repo, "__ReadMe.xlsx"),
                                   sheet = "Sheet1", col_types = "text",
                                   .name_repair = "minimal"),
                        stringsAsFactors = FALSE, check.names = FALSE)
enc_col  <- "Item encoded"
cit_col  <- "Citation (APA 7th-Annotated)"
auth_col <- "1st Author"; year_col <- "year"
readme[[enc_col]] <- as.character(readme[[enc_col]])
by_enc   <- readme[!is.na(readme[[enc_col]]) & nzchar(readme[[enc_col]]), ]
enc_pref <- sub("_.*$", "", by_enc[[enc_col]])

rows <- lapply(tsvs, function(fn) {
  base <- sub("\\.tsv$", "", fn)
  if (grepl("_", base)) {
    label     <- sub("^.*_([^_]*)$", "\\1", base)
    ident_enc <- sub("^(.*)_[^_]*$", "\\1", base)
  } else { label <- ""; ident_enc <- base }
  ident <- utils::URLdecode(ident_enc)

  url <- ""; kind <- "Other"
  if (startsWith(ident, "10.")) {
    url <- paste0("https://doi.org/", ident); kind <- "DOI"
  } else if (grepl("^PMID", ident, ignore.case = TRUE)) {
    url  <- paste0("https://pubmed.ncbi.nlm.nih.gov/", gsub("[^0-9]", "", ident), "/")
    kind <- "PubMed"
  } else if (grepl("^UMI", ident, ignore.case = TRUE)) {
    kind <- "Dissertation (ProQuest)"
  }

  hit <- which(by_enc[[enc_col]] == base)
  if (!length(hit)) hit <- which(enc_pref == ident_enc)
  cit <- auth <- yr <- ""
  if (length(hit)) {
    h <- hit[1]
    cit  <- trim(by_enc[[cit_col]][h]);  if (is.na(cit)  || cit  == "NA") cit  <- ""
    auth <- trim(by_enc[[auth_col]][h]); if (is.na(auth) || auth == "NA") auth <- ""
    yr   <- trim(by_enc[[year_col]][h]); if (is.na(yr)   || yr   == "NA") yr   <- ""
  }
  short <- if (nzchar(auth) && nzchar(yr)) sprintf("%s et al. (%s)", auth, yr)
           else if (nzchar(auth)) auth
           else if (nzchar(cit)) substr(strsplit(cit, ".", fixed = TRUE)[[1]][1], 1, 60)
           else ident

  rm <- ""
  for (cand in c(paste0(base, ".ReadMe.md"), paste0(ident_enc, ".ReadMe.md")))
    if (cand %in% mds) { rm <- cand; break }

  tab <- read.delim(file.path(pub, fn), stringsAsFactors = FALSE,
                    check.names = FALSE, quote = "\"", colClasses = "character")
  data.frame(file = fn, identifier = ident, id_type = kind, table_label = label,
             url = url, readme = rm, n_rows = nrow(tab), n_cols = ncol(tab),
             columns = paste(names(tab), collapse = "; "),
             citation = cit, citation_short = short,
             first_author = auth, year = yr,
             stringsAsFactors = FALSE)
})
manifest <- do.call(rbind, rows)
write.csv(manifest, file.path(out, "source_manifest.csv"),
          row.names = FALSE, fileEncoding = "UTF-8")
message("manifest rows: ", nrow(manifest),
        " | with citation: ", sum(nzchar(manifest$citation)))
message("DONE -> ", out)

File: __ShinyApp/build_data.py
#!/usr/bin/env python3
"""
build_data.py -- regenerate __ShinyApp/data/ from the canonical repo files.

data/ is a BUILD OUTPUT, not hand-maintained. Re-run this whenever the merge
or trait tables change, and before deploying:

    python3 __ShinyApp/build_data.py

Single source of truth = the canonical files in the repo. Requires: openpyxl
(pip install openpyxl).
"""
import os, csv, re, shutil, urllib.parse, sys
import openpyxl

APP  = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.abspath(os.path.join(APP, ".."))
OUT  = os.path.join(APP, "data")
os.makedirs(OUT, exist_ok=True)

# ---- 1. fallback copies of the two compiled long tables ---------------------
# (Primary source at runtime is GitHub; these are only a local fallback.)
shutil.copyfile(os.path.join(REPO, "__merging_volumes", "volumes_long.csv"),
                os.path.join(OUT, "volumes_long.csv"))
shutil.copyfile(os.path.join(REPO, "__merging_cellcounts", "cellcounts_long.csv"),
                os.path.join(OUT, "cellcounts_long.csv"))

# ---- 2. melt the EvoM1 trait tables -> evom1_traits_long.csv -----------------
TT = os.path.join(REPO, "____EvoM1_TraitTable")
TRAIT_FILES = {
    "dexterity_corticospinaltract.xlsx": "Dexterity & corticospinal tract",
    "corticospinaltract_etc.xlsx":       "Corticospinal tract & ecology",
    "glia_gyrification.xlsx":            "Glia, gyrification & life history",
    "interlaminar_astrocytes.xlsx":      "Interlaminar astrocytes",
}
ID_COLS = {"species_sci", "Species", "Animal", "Species Generic Name"}
SRC_SUFFIX = ["_Source", " Source", "_Ref", " Ref", "_ref", " ref"]
is_src  = lambda c: any(c.endswith(s) for s in SRC_SUFFIX)
def base_of(c):
    for s in SRC_SUFFIX:
        if c.endswith(s):
            return c[:-len(s)].strip()
    return c

rows_out = []
for fn, label in TRAIT_FILES.items():
    ws = openpyxl.load_workbook(os.path.join(TT, fn), read_only=True, data_only=True)["Sheet1"]
    data = list(ws.iter_rows(values_only=True))
    hdr = [str(h).strip() if h is not None else "" for h in data[0]]
    src_idx = {base_of(c): i for i, c in enumerate(hdr) if is_src(c)}
    sp_i = hdr.index("Species") if "Species" in hdr else 0
    for r in data[1:]:
        sp = r[sp_i]
        if sp is None or not str(sp).strip() or str(sp).strip().lower() == "none":
            continue
        sp = str(sp).strip()
        for i, c in enumerate(hdr):
            if not c or c in ID_COLS or is_src(c):
                continue
            v = r[i]
            if v is None or not str(v).strip():
                continue
            vs = str(v).strip()
            if vs.lower() in ("na", "nan", "none", "-"):
                continue
            src = label
            if c in src_idx:
                sv = r[src_idx[c]]
                if sv is not None and str(sv).strip():
                    src = str(sv).strip()
            rows_out.append((sp, c, vs, src))

seen, dedup = set(), []
for row in rows_out:
    if row not in seen:
        seen.add(row); dedup.append(row)
with open(os.path.join(OUT, "evom1_traits_long.csv"), "w", newline="", encoding="utf-8") as fh:
    w = csv.writer(fh); w.writerow(["Species", "Variable", "Value", "Source"]); w.writerows(dedup)
print("evom1 traits rows:", len(dedup))

# ---- 3. index the public source tables (NOT copied) -------------------------
# The 180 source TSVs are served straight from __Public/comparative-data on
# GitHub at runtime, so they are not duplicated into the app. We only read them
# here to record row/column counts in the manifest.
PUB = os.path.join(REPO, "__Public", "comparative-data")
tsvs = sorted(f for f in os.listdir(PUB) if f.endswith(".tsv"))
mds  = [f for f in os.listdir(PUB) if f.endswith(".ReadMe.md")]
print("source tables indexed (served from GitHub):", len(tsvs))

# ---- 4. build source_manifest.csv (filenames + citations from __ReadMe.xlsx) -
wb = openpyxl.load_workbook(os.path.join(REPO, "__ReadMe.xlsx"), read_only=True, data_only=True)
ws = wb["Sheet1"]; rr = list(ws.iter_rows(values_only=True))
hdr = [str(h).strip() if h else "" for h in rr[0]]
C = {n: hdr.index(n) for n in ["Citation (APA 7th-Annotated)", "Item encoded", "1st Author", "year"]}
def g(r, n):
    v = r[C[n]]; return str(v).strip() if v is not None else ""
by_enc, by_pref = {}, {}
for r in rr[1:]:
    enc = g(r, "Item encoded")
    if enc:
        by_enc[enc] = r
        by_pref.setdefault(enc.split("_")[0], r)

def parse(fn):
    base = fn[:-4]
    ident_enc, label = (base.rsplit("_", 1) + [""])[:2] if "_" in base else (base, "")
    ident = urllib.parse.unquote(ident_enc)
    url, kind = "", "Other"
    if ident.startswith("10."):
        url, kind = "https://doi.org/" + ident, "DOI"
    elif ident.upper().startswith("PMID"):
        url, kind = "https://pubmed.ncbi.nlm.nih.gov/%s/" % re.sub(r"[^0-9]", "", ident), "PubMed"
    elif ident.upper().startswith("UMI"):
        kind = "Dissertation (ProQuest)"
    return ident_enc, ident, label, url, kind

manifest = []
for fn in tsvs:
    ident_enc, ident, label, url, kind = parse(fn)
    base = fn[:-4]
    r = by_enc.get(base) or by_pref.get(ident_enc)
    cit = g(r, "Citation (APA 7th-Annotated)") if r else ""
    auth = g(r, "1st Author") if r else ""
    yr = g(r, "year") if r else ""
    if auth and yr: short = "%s et al. (%s)" % (auth, yr)
    elif auth:      short = auth
    elif cit:       short = cit.split(".")[0][:60]
    else:           short = ident
    rm = ""
    for cand in (base + ".ReadMe.md", ident_enc + ".ReadMe.md"):
        if cand in mds:
            rm = cand; break
    with open(os.path.join(PUB, fn), newline="", encoding="utf-8", errors="replace") as fh:
        tab = list(csv.reader(fh, delimiter="\t"))
    ncol = len(tab[0]) if tab else 0
    hdrcols = "; ".join(h.strip().strip('"') for h in (tab[0] if tab else []))
    manifest.append(dict(file=fn, identifier=ident, id_type=kind, table_label=label,
                         url=url, readme=rm, n_rows=max(0, len(tab) - 1), n_cols=ncol,
                         columns=hdrcols, citation=cit, citation_short=short,
                         first_author=auth, year=yr))
fields = ["file", "identifier", "id_type", "table_label", "url", "readme",
          "n_rows", "n_cols", "columns", "citation", "citation_short",
          "first_author", "year"]
with open(os.path.join(OUT, "source_manifest.csv"), "w", newline="", encoding="utf-8") as fh:
    w = csv.DictWriter(fh, fieldnames=fields); w.writeheader(); w.writerows(manifest)
print("manifest rows:", len(manifest),
      "| with citation:", sum(1 for m in manifest if m["citation"]))
print("DONE ->", OUT)
